### PR TITLE
[SPARK-43412][PYTHON][CONNECT] Introduce `SQL_ARROW_BATCHED_UDF` EvalType for Arrow-optimized Python UDFs

### PR DESCRIPTION
**Pull request by xinrong-meng**
### What changes were proposed in this pull request?
Introduce `SQL_ARROW_BATCHED_UDF` EvalType for Arrow-optimized Python UDFs.

An EvalType is used to uniquely identify a UDF type in PySpark.

### Why are the changes needed?
We are about to improve nested non-atomic input/output support of an Arrow-optimized Python UDF.

However, currently, it shares the same EvalType with a pickled Python UDF, but the same implementation with a Pandas UDF.

Introducing an EvalType enables isolating the changes to Arrow-optimized Python UDFs.

The PR is also a pre-requisite for registering an Arrow-optimized Python UDF.

### Does this PR introduce _any_ user-facing change?
No user-facing behavior/result changes for Arrow-optimized Python UDFs.

An `evalType`, as an attribute mainly designed for internal use, is changed as shown below:

```py
>>> udf(lambda x: str(x), useArrow=True).evalType == PythonEvalType.SQL_ARROW_BATCHED_UDF
True

# whereas

>>> udf(lambda x: str(x), useArrow=False).evalType == PythonEvalType.SQL_BATCHED_UDF
True
```

### How was this patch tested?
A new unit test `test_eval_type` and existing tests.

SPARK-40307

### CODE REVIEW

File: python/pyspark/sql/connect/udf.py [hunk 1]
```diff
@@ -253,15 +253,17 @@ def register(
             f = cast("UserDefinedFunctionLike", f)
```
Comment (xinrong-meng) [inline python/pyspark/sql/connect/udf.py:253]: Before the change proposed in that PR, registering Arrow-optimized Python UDF is buggy: the UDF registered will be executed as a non-Arrow-optimized Python UDF.
That should be fixed separately.

**Review by zhengruifeng — APPROVED** [review]

**Review by HyukjinKwon — APPROVED** [review]

**Comment by xinrong-meng** [comment]
cc @ueshin

**Comment by xinrong-meng** [comment]
Merged to master, thanks!

**Comment by xinrong-meng** [comment]
Please feel free to leave comments if any, I'll adjust them in a follow-up.

File: python/pyspark/sql/udf.py [hunk 2]
```diff
@@ -253,6 +250,7 @@ def __init__(
     def returnType(self) -> DataType:
         # This makes sure this is called after SparkContext is initialized.
         # ``_parse_datatype_string`` accesses to JVM for parsing a DDL formatted string.
+        # TODO: PythonEvalType.SQL_BATCHED_UDF
```
Comment (allisonwang-db) [inline python/pyspark/sql/udf.py:253]: @xinrong-meng just curious what is the TODO item here?

Comment (HyukjinKwon) [reply python/pyspark/sql/udf.py:253]: This is why we should have ID'ed todo .. like `TODO(SPARK-XXXXX): ...`

**Pull request closed without merging**